React 18

dameyerdave opened the pull request:
This pull request contains the following adjustments:
- Update to react v18
- Update related packages
- Resolved all vulnerabilities

Open Problem:
- All tests are running through on my local machines (`npm run test`)
- Snapshot-Tests on Github are not running without complaining even if I updated the Snapshots (because they changed according to the react v18 and related packages update)
- I also tried `act` (https://github.com/nektos/act) to check if there are issues on the Github images - but locally the Githup actions ran without any errors. 
@chaoran-chen: Could you please check why the Snapshots are not treated correctly - I have no clue by now. 

BREAKING CHANGE: Please test carefully before you merge.
Resolves #563

dameyerdave committed:
Updated to react-18

dameyerdave committed:
Updated react-bootstrap

dameyerdave committed:
[FIX] Fixed the routes to be compatible with react-router v6

dameyerdave committed:
[FIX] Fixed the useMatch method to mimic the same behavior as before react-route v6

dameyerdave committed:
[FIX] Elements in a list need a key

dameyerdave committed:
[FIX] A checkbox can only ether be controlled or not: removed the defaultChecked property if checked is given by the state.

dameyerdave committed:
[VER] Updated packages to the latest version

dameyerdave committed:
[FIX] Adjusted code to support react v18

dameyerdave committed:
[FIX] remove StrictMode

dameyerdave committed:
[CLE] Cleaned up console logs

dameyerdave committed:
[CLE] Reformatting code according to the guidelines

dameyerdave committed:
[CLE] Remove some console logs that are no longer used

dameyerdave committed:
[TEST] Update snapshots that changed because of the update of the rechart Tooltip

dameyerdave committed:
[TEST] Fixed VariantAgeDistributionChart test to run the renderer function inside act. Otherwise the setReady will never be set to true and the diagram is not rendered

dameyerdave committed:
[TEST] Rechart modules are need an additional moduleNameMapper

dameyerdave committed:
[TEST] Updated snapshots

dameyerdave committed:
[SEC] Fixed package versions to 0 vulnerabilities

dameyerdave committed:
[FIX] Link to must be a absolute path or a relative one from the current path. This is new in react-18 router implementation.

dameyerdave committed:
[TEST] Update github actions version

dameyerdave committed:
[FIX] Adjusted formatter in rechart Tooltip (this can not be the final solution!)

chaoran-chen commented:
Hi @dameyerdave, thank you very much! This is a huge contribution! Could you please just resolve the merge conflicts?

@fengelniederhammer @JonasKellerer Could you please review this PR and help with the open problems? I won't have time this and next week as I'll be traveling next week.

dameyerdave committed:
Merge branch 'develop' into react-18

dameyerdave commented:
Sure thing - merge done

dameyerdave commented:
Give me a bit to clean up the code now...

dameyerdave committed:
fix(packages): Seams that resize-observer-browser is no longer needed

dameyerdave committed:
fix: removed not used declatations

Especially the Fragments that are now references as React.Fragment

dameyerdave committed:
fix: Fixes the declaration of DefaultTooltipContent

The declaration has changes for the later version of recharts.

dameyerdave committed:
fix(validateDOMnesting): <div> cannot appear as a descendant of <p>

if using Typography inside a component producing a `<p>` this will output a validateDOMnesting warning

SOLUTION: https://stackoverflow.com/questions/41928567/div-cannot-appear-as-a-descendant-of-p

dameyerdave committed:
fix(key): Components need a key

Also the svg path element needs it

dameyerdave committed:
fix(test): should await for userEvent.selectOptions

According to the docs userEvent.selectOptions is an asynchronous function and we should wait for it before checking the expected values.

REFERENCE: https://testing-library.com/docs/user-event/utility

dameyerdave committed:
refactor(update): update package versions

dameyerdave committed:
Merge branch 'develop' into react-18

dameyerdave committed:
fix(eslint): no type any

dameyerdave commented:
I fixed merge conflicts and updated all the packages. I also fixed some other bugs in this PR. I hope you can review it and apply it soon to not force more merge conflicts.

Many thanks

fengelniederhammer commented on package.json:
```diff
@@ -145,6 +158,7 @@
     "moduleNameMapper": {
       "^react-native$": "react-native-web",
       "^.+\\.module\\.(css|sass|scss)$": "identity-obj-proxy",
+      "(lib-vendor/.*)": "victory-vendor/$1",
```
Why do we actually need this line?

dameyerdave replied:
Thus I needed to add because of an error in one of the d3 packages. Comment it out and try to npm i and you will see what fails.

fengelniederhammer commented on src/maps/Map.tsx:
```diff
@@ -42,6 +42,8 @@ export interface VectorMapProps {
+  children?: React.ReactNode;
 }
 
 const VectorMap: React.FC<VectorMapProps> = ({
```
You could also make this a `React.FC<PropsWithChildren<VectorMapProps>>` instead of adding the children to the Props interface. But doesn't change much.

dameyerdave replied:
Yes, but as you mention: I wouldn't change much...

fengelniederhammer commented on package.json:
```diff
@@ -3,79 +3,92 @@
+    "@types/react-router-dom": "^5.3.3",
+    "@types/react-simple-maps": "^3.0.0",
+    "@types/react-test-renderer": "^18.0.0",
+    "@types/recharts": "^1.8.24",
```
We removed that in https://github.com/GenSpectrum/cov-spectrum-website/commit/f9c50b81781d316895d05c4bed2ab5b7a5431cb5. Did you bring it back intentionally? I think the types were in the wrong version, and are not necessary anymore since recharts 2.0

dameyerdave replied:
I didn't do this intentionally. This is just a merge conflict I solved with the best of my knowledge and belief. If they are not needed anymore you can remove them again.

fengelniederhammer commented on src/components/SvgVennDiagram.tsx:
```diff
@@ -383,7 +384,7 @@ export const SvgVennDiagram = ({ selectors, domain, numberOfvariants }: Props) =
                   max={100}
                   step={5}
-                  onChange={value => setMinProportion(value / 100)}
+                  onChange={value => setMinProportion((value as number) / 100)}
```
Why cast here? Can't we simply declare `(value: number) => ...`?

fengelniederhammer commented:
We tried to investigate what goes wrong in the snapshot tests on Github, but we don't have an idea yet.

chaoran-chen committed:
test: fix snapshot tests

For unknown reasons, the CI system were previously producing different snapshots than the devices of the developers.

chaoran-chen commented:
I don't know what caused the discrepancy between the CI server and our local devices. The CI executes the `useEffects` before creating the snapshots, but our local devices do not. I think that the CI's behavior is what we want and the `await act(async () => {});` statements that we have in the tests are intended to make exactly that happen.

For me, writing `await act(async () => {});` twice in the failing tests is doing the trick. Could you please check whether you are now also getting the same snapshots?

chaoran-chen commented on src/components/VariantSearchField.tsx:
```diff
@@ -1,8 +1,8 @@
+import { components, CSSObjectWithLabel, InputActionMeta, StylesConfig } from 'react-select';
 import { isValidAAMutation, isValidABNotation, isValidNspNotation } from '../helpers/aa-mutation';
-import { CSSPseudos } from 'styled-components';
+// import { CSSPseudos } from 'styled-components';
```
Remove commented-out code

chaoran-chen committed:
refactor: minor cleanups

chaoran-chen committed:
Merge branch 'develop' into react-18

chaoran-chen reviewed (approved):
I think that this works well and is safe to be merged. Thank you, @dameyerdave for implementing, and thanks for reviewing, @fengelniederhammer!

chaoran-chen commented:
Oh, there was an error but I never noticed :( See be0ec6cacfd986f29e7d1cadce4d48c67dfd2e88
